File: leetcode/medium/931-min-falling-path/rec-dp.py
"""
  @author Vortexx2
  Problem 931 - Minimum Falling Path

  Runtime - 315 ms
  Memory Usage - 22.8 MB
"""
from typing import List


class Solution:
  def minFallingPathSum(self, matrix: List[List[int]]) -> int:

    dp = {}

    def helper(row: int, col: int) -> int:

      if col == -1 or col == len(matrix):
        return float('inf')

      if row == len(matrix) - 1:
        return matrix[row][col]

      if (row, col) in dp:
        return dp[(row, col)]

      dp[(row, col)] = matrix[row][col] + min(helper(row + 1, col - 1),
                                              helper(row + 1, col),
                                              helper(row + 1, col + 1))

      return dp[(row, col)]

    minVal = helper(0, 0)
    for i in range(1, len(matrix)):
      minVal = min(minVal, helper(0, i))

    return minVal

File: leetcode/medium/931-min-falling-path/rec-dp.cpp
/**
 * @file rec-dp.cpp
 * @author Vortexx2
 * @brief Problem 931 - Minimum Falling Path
 * @approach - Using a recursive DP approach
 * @date 29-07-2022
 *
 * Runtime - 126 ms
 * Memory Usage - 16.2 MB
 */
#include <climits>
#include <iostream>
#include <map>
#include <vector>

using namespace std;

class Solution {
 public:
  map<pair<int, int>, int> dp;
  int minFallingPathSum(vector<vector<int>>& matrix) {
    int minVal = INT_MAX;
    for (int i = 0; i < matrix.size(); i++)
      minVal = min(minVal, helper(matrix, 0, i));

    return minVal;
  }

  int helper(vector<vector<int>>& matrix, int row, int col) {
    if (col == -1) return INT_MAX;

    if (col == matrix.size()) return INT_MAX;

    if (row == matrix.size() - 1) return matrix[row][col];

    if (dp.find({row, col}) != dp.end()) return dp[{row, col}];

    dp[{row, col}] =
        matrix[row][col] + min(helper(matrix, row + 1, col - 1),
                               min(helper(matrix, row + 1, col),
                                   helper(matrix, row + 1, col + 1)));

    return dp[{row, col}];
  }
};